synapse/lib/layer.py

```diff
@@ -1204,6 +1213,182 @@ async def __anit__(self, layrinfo, dirn, nexsroot=None, allow_upstream=True, map
+        abrv = self.tagabrv.bytsToAbrv(tagname.encode())
+        abrv += self.getPropAbrv(formname, None)
+        if not self.layrslab.hasdup(abrv, buid, db=self.bytag):
+            yield ('NoTagIndex', {'buid': buid, 'form': formname, 'tag': tagname, 'valu': tagvalu})
+
+    def _testDelTagIndx(self, buid, form, tag):
+        formabrv = self.setPropAbrv(form, None)
+        tagabrv = self.tagabrv.bytsToAbrv(tag.encode())
+        self.layrslab.delete(tagabrv + formabrv, buid, db=self.bytag)
+
+    def _testDelPropIndx(self, buid, form, prop):
+        sode = self._getStorNode(buid)
+        storvalu, stortype = sode['props'][prop]
+
+        abrv = self.setPropAbrv(form, prop)
+        for indx in self.stortypes[stortype].indx(storvalu):
+            self.layrslab.delete(abrv + indx, buid, db=self.byprop)
+
+    def _testDelTagStor(self, buid, form, tag):
+        sode = self._getStorNode(buid)
+        sode['tags'].pop(tag, None)
+        self.setSodeDirty(buid, sode, form)
+
+    def _testDelPropStor(self, buid, form, prop):
+        sode = self._getStorNode(buid)
+        sode['props'].pop(prop, None)
+        self.setSodeDirty(buid, sode, form)
+
+    def _testAddPropIndx(self, buid, form, prop, valu):
+        sode = self._getStorNode(buid)
+        storvalu, stortype = sode['props'][prop]
+
+        abrv = self.setPropAbrv(form, prop)
+        for indx in self.stortypes[stortype].indx(valu):
+            self.layrslab.put(abrv + indx, buid, db=self.byprop)
+
+    async def verify(self, buids=True, tags=True, props=True):
+        if buids:
+            async for error in self.verifyAllBuids():
+                yield error
+
+        if tags:
+            async for error in self.verifyAllTags():
+                yield error
+
+        if props:
+            async for error in self.verifyAllProps():
+                yield error
+
+    async def verifyAllBuids(self):
+        async for buid, sode in self.getStorNodes():
+            async for error in self.verifyByBuid(buid, sode=sode):
+                yield error
+
+    async def verifyAllTags(self):
+        for name in self.tagabrv.names():
+            async for error in self.verifyByTag(name):
+                yield error
+
+    async def verifyAllProps(self):
+        for form, prop in self.getFormProps():
+            async for error in self.verifyByProp(form, prop):
+                yield error
+
+    async def verifyByTag(self, tag):
+        tagabrv = self.tagabrv.bytsToAbrv(tag.encode())
+        for lkey, buid in self.layrslab.scanByPref(tagabrv, db=self.bytag):
+
+            await asyncio.sleep(0)
+
+            sode = self._getStorNode(buid)
+            if sode is None:
+                yield ('NoBuidForTagIndex', {'buid': buid, 'tag': tag})
+                continue
+
+            tags = sode.get('tags')
+            if tags is None:
+                yield ('NoTagForTagIndex', {'buid': buid, 'tag': tag})
```

Comment:
This error differs from the next one an represent different conditions?

Reply:
Not really, they both represent the non-presence of a tag value it's just that we need 2 cases to account for their being ( or not being ) any others.